### ISSUE TITLE
Spatial join fails if column has all empty geometries

### ISSUE THREAD

**Issue by ian-r-rose** [geopandas/geopandas#1315]
I ran into this edge case with `geopandas.sjoin`. If the left GeoDataFrame in a join has all empty geometries, then a spatial index is not generated, leading to an `AttributeError` further in the function.

Here is a minimal example:

```python
import geopandas
import shapely

df1 = geopandas.GeoDataFrame(geometry=[shapely.geometry.Point(0,0)])
df2 = geopandas.GeoDataFrame(geometry=[shapely.geometry.GeometryCollection()])

# geopandas.sjoin(df1, df2) # works
geopandas.sjoin(df2, df1) # fails
```

Which produces:

```pytb
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-17-e670cb379738> in <module>
      7 # geopandas.sjoin(df1, df2) # works
      8 # df1.sindex
----> 9 geopandas.sjoin(df2, df1) # fails

/opt/conda/lib/python3.7/site-packages/geopandas/tools/sjoin.py in sjoin(left_df, right_df, how, op, lsuffix, rsuffix)
    117         # tree_idx_df == 'left'
    118         idxmatch = right_df.geometry.apply(lambda x: x.bounds).apply(
--> 119             lambda x: list(tree_idx.intersection(x)) if not x == () else []
    120         )
    121         idxmatch = idxmatch[idxmatch.apply(len) > 0]

/opt/conda/lib/python3.7/site-packages/pandas/core/series.py in apply(self, func, convert_dtype, args, **kwds)
   4043             else:
   4044                 values = self.astype(object).values
-> 4045                 mapped = lib.map_infer(values, f, convert=convert_dtype)
   4046 
   4047         if len(mapped) and isinstance(mapped[0], Series):

pandas/_libs/lib.pyx in pandas._libs.lib.map_infer()

/opt/conda/lib/python3.7/site-packages/geopandas/tools/sjoin.py in <lambda>(x)
    117         # tree_idx_df == 'left'
    118         idxmatch = right_df.geometry.apply(lambda x: x.bounds).apply(
--> 119             lambda x: list(tree_idx.intersection(x)) if not x == () else []
    120         )
    121         idxmatch = idxmatch[idxmatch.apply(len) > 0]

AttributeError: 'NoneType' object has no attribute 'intersection'
```

A workaround is to manually trigger index creation in the other GeoDataFrame:
```python
import geopandas
import shapely

df1 = geopandas.GeoDataFrame(geometry=[shapely.geometry.Point(0,0)])
df2 = geopandas.GeoDataFrame(geometry=[shapely.geometry.GeometryCollection()])

df1.sindex
geopandas.sjoin(df2, df1) # works now
```

I am running this with geopandas 0.7.0. Similar to, but distinct from #251, #944, as far as I can tell.

**Comment by martinfleis** [comment]
Thanks for the report. It is not related to those linked issues, but it is an unexpected corner case. It would require one check after which we should either raise an error that there is nothing to join to (no valid geometry) or return original gdf. What would you expect to happen in this case?

**Comment by ian-r-rose** [comment]
My expectation is that a spatial join on a column full of empty geometries would produce no matches on the predicate op. So rather than erroring, or returning the original gdf, it would return an empty gdf.

My reasoning is that this would be less surprising in the case that a gdf has a few non-empty rows. If it is mostly-empty, the spatial join will join on the few non-empty ones. If, then, those non-empty rows are dropped, I would expect the resulting join would drop correspondingly.

To make that more concrete, consider the following:
```python
import geopandas
import shapely

df1 = geopandas.GeoDataFrame(geometry=[shapely.geometry.Point(0,0)])
df2 = geopandas.GeoDataFrame(geometry=[
    shapely.geometry.Point(0,0),
    shapely.geometry.Point(0,0),
    shapely.geometry.GeometryCollection(),
])
```

Currently, the following join
```python
geopandas.sjoin(df2, df1)
```
produces
<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>geometry</th>      <th>index_right</th>    </tr>  </thead>  <tbody>    <tr>      <th>0</th>      <td>POINT (0.00000 0.00000)</td>      <td>0</td>    </tr>    <tr>      <th>1</th>      <td>POINT (0.00000 0.00000)</td>      <td>0</td>    </tr>  </tbody></table>

If I drop the first row
```python
geopandas.sjoin(df2.iloc[1:], df1)
```
it produces:
<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>geometry</th>      <th>index_right</th>    </tr>  </thead>  <tbody>    <tr>      <th>1</th>      <td>POINT (0.00000 0.00000)</td>      <td>0</td>    </tr>  </tbody></table>


If I drop all but the last row 
```python
geopandas.sjoin(df2.iloc[2:], df1)
```
it errors, but for consistency with the above, I would expect:

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>geometry</th>      <th>index_right</th>    </tr>  </thead>  <tbody>     </tbody></table>

**Comment by martinfleis** [comment]
That makes sense. Just keep in mind that it would be different for `how=left`. Would you like to make a PR fixing this?

**Comment by ian-r-rose** [comment]
Sure, I would be happy to make a PR. I suspect that a fix may also run into #1307, anything I should keep in mind?

**Comment by martinfleis** [comment]
Fixing both would be great, thanks! I can't think of anything else than making sure each of the `how` options is taken care of at the moment.
